src/datasources/prices-api/coingecko-api.service.ts

```diff
@@ -146,19 +155,99 @@ export class CoingeckoApi implements IPricesApi {
     }
   }
 
-  async registerNotFoundTokenPrice(args: {
+  private _mapProviderError(error: any): string {
+    const errorCode = error?.status?.error_code;
+    return errorCode ? ` [status: ${errorCode}]` : '';
+  }
+
+  /**
+   * For an array of token addresses, gets the prices available from cache.
+   */
+  private async _getTokenPricesFromCache(args: {
     chainName: string;
-    tokenAddress: string;
+    tokenAddresses: string[];
     fiatCode: string;
-  }): Promise<void> {
-    const cacheDir = CacheRouter.getTokenPriceCacheDir(args);
-    const expireTimeSeconds = this._getRandomNotFoundTokenPriceTtl();
-    return this.cacheService.expire(cacheDir.key, expireTimeSeconds);
+  }): Promise<AssetPrice[]> {
+    const result = await Promise.all(
+      args.tokenAddresses.map(async (tokenAddress) => {
+        const cacheDir = CacheRouter.getTokenPriceCacheDir({
+          ...args,
+          tokenAddress,
+        });
+        const cached = await this.cacheService.get(cacheDir);
+        const { key, field } = cacheDir;
+        if (cached != null) {
+          this.loggingService.debug({ type: 'cache_hit', key, field });
+          return JSON.parse(cached);
+        }
+        this.loggingService.debug({ type: 'cache_miss', key, field });
+        return null;
+      }),
+    );
+
+    return result.filter((i) => i !== null);
   }
 
-  private _mapProviderError(error: any): string {
-    const errorCode = error?.status?.error_code;
-    return errorCode ? ` [status: ${errorCode}]` : '';
+  /**
+   * For an array of token addresses, gets the prices available from the Coingecko API.
+   * Stores both retrieved prices and not-found prices in cache.
+   */
+  private async _getTokenPricesFromNetwork(args: {
+    chainName: string;
+    tokenAddresses: string[];
+    fiatCode: string;
+  }): Promise<AssetPrice[]> {
+    const prices = await this._requestPricesFromNetwork({
+      ...args,
+      tokenAddresses: args.tokenAddresses.slice(0, CoingeckoApi.maxBatchSize),
+    });
+
+    return Promise.all(
+      args.tokenAddresses.map(async (tokenAddress) => {
+        const validPrice = prices.data[tokenAddress]?.[args.fiatCode];
+        const price: AssetPrice = validPrice
+          ? { [tokenAddress]: { [args.fiatCode]: validPrice } }
+          : { [tokenAddress]: { [args.fiatCode]: null } };
+        await this.cacheService.set(
+          CacheRouter.getTokenPriceCacheDir({ ...args, tokenAddress }),
+          JSON.stringify(price),
```

Comment:
To clarify: a `price` is a single token price correct?

The `AssetPrice` supports multiple tokens but we have changed that to return an array of `AssetPrice` where each `AssetPrice` is a single token correct?

Reply:
Yes, that's correct. I can include a comment in the code to clarify this part. `AssertPrice` can hold multiple tokens (it models the response from CoinGecko). I thought it would be easier for the clients of this class to receive an array of items instead of an object of 0..N keys so that complexity is hidden within the class.

Another option would be having different entities for the datasource part (CoinGecko mappings) and the domain. I thought this would add complexity but I'm open if you think it's better.

Reply:
GIven the nature of this feature and how we are adapting this solution over time I would say that the way you have it is also ok for its purpose 👍